fabric-fluid-v1/src/main/java/net/fabricmc/fabric/api/fluids/v1/math/Drops.java

```diff
@@ -0,0 +1,73 @@
+package net.fabricmc.fabric.api.fluids.v1.math;
+
+import net.minecraft.nbt.CompoundTag;
+
+/**
+ * Drops are the standard unit of fluid for the fluid api.
+ * 1/{@link #BUCKET} is the smallest amount of fluid that can be described with the Drops system.
+ * Stored in a long, at one bucket per tick, would take around 11.6 million years to overflow!
+ */
+public class Drops {
+	/**
+	 * because of constant inlining use {@link #getBucket()} and {@link #getBuckets(long)}.
+	 */
+	private static final long BUCKET = 2520L /*this number is divisible by all numbers from 1 to 9 inclusive.*/;
```

Comment:
1 to 10 inclusive, but this still doesn't cover the TCon usecase and many others; I still think a hardcoded number is a Not So Good Approach, and knowing things we'll be stuck with it the second it gets merged.

Reply:
what's the TCon use case?
I'd prefer a fixed denominator system over Fractions, and I made sure the number can be changed, hence the to and from tag methods. I know as well that we'll never agree on one number 😉

Reply:
Metal liquids. If we assume a block is 1 bucket, an ingot is 1/9 bucket and a nugget is 1/81 bucket. In the land of forges, the mod is forced to treat metal blocks as something else than 1000mB per block, say, 1296mB per block, which creates a kind of mental conflict.

Also, there's still the question of intuitiveness: milli-units are very intuitive, while Drops will not be as intuitive. Fractions can go either way depending on how mods use them, as such I think they are more Fabric-like as they offer far more freedom.

Reply:
- fractions waste almost all of the data they're given (since most fractions can be simplified)
- they're also objects and not primitives
- imagine the following case: pipes with each 5 sides (1 input) connected:
  - it only takes 27 pipes before you get a loss of precision
  - now tell me, what's worse, having 2519/2520 buckets or (longmax-1)/longmax :concernedtater:, long max, since there's basically no way of getting that amount of fluid again
  - along with that, people wont naturally account for loss of precision in a fraction system since the unit is microscopic (longmax), when the unit is relatively thicc, people understand that loss is something to account for, and will do so
 - and if u make everyone account for precision, then you might as well just use a fixed denominator system

this is why fraction bad, fixed denominator good, also it's not hardcoded irritatered

Reply:
🚲🏠

Reply:
fractions is pretty future proof, if we are changing the dominator anytime, what's the difference in using fractions. heck, if you change the dominator you will be breaking stuff if you are saving a long

Reply:
https://github.com/FabricMC/fabric/pull/657#discussion_r434746513

fractions have too many more problems than something like this

Reply:
you can reasonably ish get it back, much easier than 1/long_max atleast. And why not? and it's not my only deciding factor either

Reply:
you *will* have loss of precision with fractions, and it'll have worse consequences over a fixed denominator system, in a fix denominator system everyone knows and understands there will be loss of precision, you can't say the same for fractions, even though it's fairly easy for it to happen. If you have to account for loss of precision in fractions you might as well just use a fixed denominator system anyways.

Reply:
where is the lost of precision came from

Reply:
https://github.com/FabricMC/fabric/pull/657#discussion_r434746513 smh

Reply:
it's not duplicating anything, you are just splitting the fluid up and up

Reply:
remember that the denominator is a long, and it's 2^63 (signed), so splitting an odd fraction 64 times will result in loss of precision

Reply:
and to split a fraction 64 times is pretty common, like in pipes, I used 5 as an example just to show the extreme, but it's not uncommon to have that either

Reply:
hmm that sounds reasonable, ~~BigInteger~~ ok no

Reply:
concernedtater

Reply:
Explain to me how you're gonna tell all the players that have been using `mB` for nearly one decade that they suddenly have 37/64 water in their tanks.

Fractions are cool from a theoretical standpoint, and fun to play with and use, but I just don't see a reason to innovate in that regard. It makes conversion harder for the end user unless you write a fancy way of displaying them as decimals, at which point you... just entirely defeated the reason for fractions. It'd be easier to build a fractional system on top of a raw measure system than the other way around.

Reply:
Forge does not determine the standard, we do not have to use base 1000 units

Reply:
That's perfectly reasonable, I used millibuckets as a common implementation of decimal fluid values. I am not opposed to a fractional API in itself, but a decimal API simply makes things easier with less work in my opinion.

Reply:
his point is that the community is more used to it, anyways it's not like a base 10 number is inherently superior. As an American I'm definitely not biased in any way

Reply:
That would be correct. Saying you have 2520 drops may at first seem less intuitive than 1000 millibuckets, but both are arbitrary values, the difference being one is base 10, and the other is not.

I would usually consider a base 10 system to be cleaner, however, dealing with fluids, having a high divisibility is useful and offsets the reduced clarity simply because of its technical advantages. Migrating from 1000mB to 2520d is easier than 1000mB to 1/1 or whichever fraction was chosen to represent a whole bucket, once you start performing operations on it.

Reply:
the "minecraft fluid api" is in 1/1, but it doesn't work, atleast not how you'd expect it to

Reply:
I blame yarn

Reply:
2520 is also a weird value, considering 1620 is used for existing Fabric fluid libraries. (also fraction good and best in game and litre/"millibucket" bad and worst)

Reply:
My issue with this value is some mod will FORCE their value by mixins, asm witchcraft, reflection or touching it in preLaunch

Reply:
I'd say that's a non-existent problem, there's no reason to account for people mixing into the api regardless. And even if that does happen the api can deal with it, since the value is not hardcoded and I made sure that the value would not get constant inlined

Reply:
If someone goes out of their way to hack the API, what gives?

Any Fabric API can be broken and changed in those ways, yet they're still let through. Simply put, intentional breaking changes by a third party are not exclusive to this API and should not determine whether it goes through or not. Other points are perfectly reasonable, but I see  that as a flawed argument.

Reply:
since I guess we're having this damn argument again, I have to acknowledge that 1620 is my greatest secret shame, because I could have sworn I'd tested it for divisibility by 8 but hadn't, and it's not. Eight was one of the most common requests for a divisor there was. If we want divisibility by 2, 3, 4, 5, 6, 8, 9, 10, 12, 27, and 81, each of which is a very reasonable number to want divisibility for, the best option for a denom is 3240. Either that or fractions to mask 81000 as a denominator, since that's divisible by 2, 3, 4, 5, 6, 8, 9, 10, 12, 27, 81, 100, and 1000.